lib/util.c

```diff
@@ -361,69 +412,244 @@ idx_1st_upper_bound(const double *values, size_t n_values, double query)
+        min_query = idx / self->query_multiplier;
+        /* move ptr to the first upper bound of min_query in elements */
+        while (ptr < stop && *ptr < min_query) {
+            ptr++;
+        }
+        /* The query range of [A, B) maps to `idx` where
+         * A = idx/query_multiplier and B = (idx + 1)/query_multiplier).
+         * Want `lookup[idx]` to point to the first upper bound of A in the elements.
+         */
+        self->lookups[idx] = (unsigned) (ptr - elements);
+    }
+out:
+    return ret;
 }
 
-/* Inputs:
- *   - `lengths` is an array of the real number lengths of `num_intervals`
- *      consecutive half-open intervals on the real number line start at zero,
- *      i.e. given lengths L_i, the intervals are:
- *          [0.0, L_0), [L_0, L_0 + L_1), [L_0 + L_1, L_0 + L_1 + L_2), ...
- *   - `x` is a real number to find in one of the intervals or after all of them
- * Returns:
- *   - the index (starting at zero) of the interval in which `x` was found
- *   - `num_intervals` if `x` is at or past the end of the last interval
- *     (the sum of all the lengths)
- *   - zero if `x` is negative
- *   - `num_interval` if `x` is NaN
- * Pre-conditions:
- *   - `lengths` must point to `num_interval` non-negative non-NaN doubles
- * Post-condition:
- *   - returned value is less than or equal to `num_intervals` (and non-negative)
- *
+/* Returns the least power of 2 greater than or equal to `x` for positive numbers,
+ * otherwise returns zero.
  */
-static inline size_t
-positive_interval_select(double x, size_t num_intervals, const double *lengths)
+static double
+higher_power_of_2(double x)
 {
-    size_t i = 0;
-    double sum = 0.0;
-    while (i < num_intervals) {
-        assert(lengths[i] >= 0.0);
-        sum += lengths[i];
-        if (x < sum) {
-            break;
-        }
-        i++;
+    assert(x >= 0);
+    if (x <= 0) {
+        return 0.0;
     }
-    return i;
+    double floor = exp2(logb(x));
+    return (floor < x ? floor * 2.0 : floor);
 }
 
-/* Convenient helper function to use with random variant `u` and array of probabilities
- * that approximately sum to one. Note the function does not actually use the last
- * probability because its correct value is implied by all the previous probabilities.
+/* PRE-CONDITIONS:
+ *     1) `elements` must point to array of doubles starting at exactly `0.0`
+ *     2) `elements` must be non-empty non-decreasing and with no NaN
  */
-size_t
-probability_list_select(double u, size_t num_probs, double const *probs)
+int
+fast_search_alloc(fast_search_t *self, const double *elements, size_t n_elements)
 {
-    return (num_probs > 0 ? positive_interval_select(u, num_probs - 1, probs) : 0);
+    int ret = 0;
+    double max_element;
+
+    const size_t max_input_size = 1LL << (DBL_MANT_DIG - 1); // 4096 terabytes
+
+    memset(self, 0, sizeof(*self));
+
+    if (!valid_sorted_nonempty_array(elements, n_elements)) {
```

Comment:
Maybe this has already been discussed - but should these parameter checks not be assertions? If they are driven by end-user input I can see that returning an error would be good. But if we are in control of the parameters then they should be assertions?

Reply:
My hunch it hasn't really been discussed. It's only recently there is a distinction between assert() vs tsk_bug_assert(). Thus not a lot of discussion yet about when one vs the other should be used.

With the introduction of tsk_bug_assert() vs assert(), me learning about the excellent API structure and layers in msprime, and other observations I have made, I've been thinking it would be useful for me to collect many ideas relating to this into a descriptive non-prescriptive doc for many of us to look at. My hope it will clarify the right trade-offs to make in cases such as in this one.

Maybe I can use this case as an example for how it can be improved.